Sources/FoundationEssentials/AttributedString/AttributedString+IndexTracking.swift

```diff
@@ -0,0 +1,168 @@
+internal import _FoundationCollections
+#endif
+
+// MARK: - Internal Index Updating
+
+extension AttributedString.Guts {
+    func _prepareTrackedIndicesUpdate(mutationRange: Range<BigString.Index>) {
+        // Move any range endpoints inside of the mutation range to outside of the mutation range since a range should never end up splitting a mutation
+        for idx in 0 ..< trackedRanges.count {
+            let lowerBoundWithinMutation = trackedRanges[idx].lowerBound > mutationRange.lowerBound && trackedRanges[idx].lowerBound < mutationRange.upperBound
+            let upperBoundWithinMutation = trackedRanges[idx].upperBound > mutationRange.lowerBound && trackedRanges[idx].upperBound < mutationRange.upperBound
+            switch (lowerBoundWithinMutation, upperBoundWithinMutation) {
+            case (true, true):
+                // Range is fully within mutation, collapse it to the start of the mutation
+                trackedRanges[idx] = Range(uncheckedBounds: (mutationRange.lowerBound, mutationRange.lowerBound))
+            case (true, false):
+                // Range starts within mutation but extends beyond mutation - remove portion within mutation
+                trackedRanges[idx] = Range(uncheckedBounds: (mutationRange.upperBound, trackedRanges[idx].upperBound))
+            case (false, true):
+                // Range starts before mutation but extends into mutation - remove portion within mutation
+                trackedRanges[idx] = Range(uncheckedBounds: (trackedRanges[idx].lowerBound, mutationRange.lowerBound))
+            case (false, false):
+                // Neither endpoint of range is within mutation, leave as-is
+                break
+            }
+        }
+    }
+    
+    func _finalizeTrackedIndicesUpdate(mutationStartOffset: Int, isInsertion: Bool, utf8LengthDelta: Int) {
+        // Update indices to point to the correct offsets based on the mutation deltas
+        for idx in 0 ..< trackedRanges.count {
+            var lowerBound = trackedRanges[idx].lowerBound
+            var upperBound = trackedRanges[idx].upperBound
+            
+            // Shift the lower bound if either:
+            //      A) The lower bound is greater than the start of the mutation (meaning it must be after the mutation due to the prepare step)
+            //      B) The lower bound is equal to the start of the mutation, but the mutation is an insertion (meaning the text is inserted before the start offset)
+            if lowerBound.utf8Offset > mutationStartOffset || (lowerBound.utf8Offset == mutationStartOffset && isInsertion), utf8LengthDelta != 0 {
+                lowerBound = string.utf8.index(string.startIndex, offsetBy: lowerBound.utf8Offset + utf8LengthDelta)
+            } else {
+                // Form new indices even if the offsets don't change to ensure the indices are valid in the newly-mutated rope
+                string.utf8.formIndex(&lowerBound, offsetBy: 0)
```

Comment:
Is this effectively the same as `string.utf8.startIndex`? and if so is there a reason for choosing one over the other?

Reply:
Not quite - `lowerBound` may not be the start index of the `AttributedString`, we just know that the raw UTF-8 offset doesn't need to change based on the mutation. This is just a way to "regenerate" a new `BigString.Index` with a valid rope path based on the existing offset. That existing offset may be at the start of the string or somewhere in the middle, but it doesn't change - we just want to make sure that any additional information in the index about the tree structure gets updated too based on the new tree layout.